book/src/raw-block.md

```diff
@@ -11,22 +11,62 @@ A *mounted (file) volume* is volume that will be mounted using a specified file
 
-The Kubernetes CSI tooling support for block volumes is still alpha.
+## Implementing Raw Block Volume Support
 
-## Usage
+CSI doesn't provide a capability query for block volumes, so COs will simply pass through requests for
+block volume creation to CSI plugins, and plugins are allowed to fail with the `InvalidArgument` GRPC
+error code if they don't support block volumes. Kubernetes doesn't make any assumptions about which CSI
+plugins support blocks and which don't, so users have to know if any given storage class is capable of
+creating block volumes.
+
+The difference between a request for a mounted (file) volume and a block volume is the `VolumeCapabilities`
+field of the request. Note that this field is an array and the created volume must support ALL of the
+capabilities requested, or else return an error. If the `AccessType` method of a `VolumeCapability`
+`VolumeCapability_Block`, then the capability is requesting a raw block volume. Unlike mount volumes, block
+volumes don't have any specific capabilities that need to be validated, although access modes still
+apply.
+
+Block volumes are much more likely to support multi-node flavors of `VolumeCapability_AccessMode_Mode`
+than mount volumes, because there's no filesystem state stored on the node side that creates any technical
+impediments to multi-attaching block volumes. There may still be good reasons to prevent multi-attaching
+block volumes, and there may be implementations that are not capable of supporting multi-attach, but it's
+generally wrong to use exactly the same set of access modes for block as for mount volumes.
+
+CSI plugins that support both mount and block volumes must be sure to check the capabilities of all CSI RPC
+requests and ensure that the capability of the request matches the capability of the volume, to avoid trying
+to do file-system-related things to block volumes and block-related things to file system volumes.
+
+Also, CSI plugins that implement the optional `GetCapacity()` RPC should note that that RPC includes
+capabilities too, and if the capacity for mount volules is not the same as the capacity for block
+volumes, that needs to be handled in the implementation of that RPC.
+
+Q: Can CSI plugins support only block volumes and not mount volumes?
+A: Yes! This is just the reverse case of supporting mount volumes only. Plugins may return `InvalidArgument`
+for any creation request with an `AccessType` of `VolumeCapability_Mount`.
+
+## Differences Between Block and Mount Volumes
+
+The main difference between block volumes and mount volumes is the expected result of the `NodePublish()`.
+For mount volumes, the CO expects the result to be a mounted directory, at `TargetPath`. For block volumes,
+the CO expect there to be a device file at `TargetPath`. The device file can by a bind-mounted device from
+the hosts `/dev` filesystem, or it can be a device node created at that location using `mknod()`.
```

Comment:
This may also imply that your driver needs to mount a "/dev" hostpath

Reply:
I don't know if we really need to mention it here since it's not really specific to block volumes, but the driver shouldn't assume that it's the only driver running on that node (and therefore owns all the devices).

Reply:
I'm not sure these are raw-block-specific issues. Nearly all CSI drivers of any type will need to have a hostpath mount of /dev or even /. Maybe this advice belongs in a more general section on tips for writing CSI plugins.